docs/asciidoc/faq.adoc

```diff
@@ -205,5 +226,113 @@ processor.publishOn(scheduler1)
          .doOnNext(i -> processNext(i))
          .subscribe();
 ----
+====
+
+== What Is a Good Pattern for Contextual Logging? (MDC)
+As described in section <<context>>, `ThreadLocal` does not make much sense in Reactive Streams, where several streams could run steps on a common thread.
+Instead of using `ThreadLocal`, Reactor offers the `Context`, a map-like object that is tied to a `Subscription` and not a `Thread`.
+
+One recurring use of `ThreadLocal` in Java is for logging frameworks to store variables that are used in their logging pattern, generally by way of a `Map` called the MDC ("Mapped Diagnostic Context").
+
+Unfortunately, this pattern is hard to opt out of and hard to apply transparently from a reactive library perspective.
+This entry of the FAQ offers a possible intermediate solution when one wants to perform log statements in relation to events in the stream (`onNext`, `onError`, and `onComplete`).
+
+The most straightforward way to perform such logging and get contextual information from the `Context` to the MDC is to wrap logging statements in a `doOnEach` operator with a little bit of boilerplate code.
+This boilerplate depends on both the logging framework/abstraction of your choice and the information you want to put in the MDC, so it has to be in your codebase.
+
+The following is an example of such a helper function around a single MDC variable and focused on logging `onNext` events, using Java 9 enhanced `Optional` API:
+
+====
+[source,java]
+----
+public static <T> Consumer<Signal<T>> logOnNext(Consumer<T> logStatement) {
+	return signal -> {
+		if (!signal.isOnNext()) return; <1>
+		Optional<String> toPutInMdc = signal.getContext().getOrEmpty("CONTEXT_KEY"); <2>
+
+		toPutInMdc.ifPresentOrElse(tpim -> {
```

Comment:
Typo? Should this be just `ifPresent`?

Reply:
nope, using the "enhanced Optional API" from Java 9 to cover the case where there's no interesting value in the `Context` (second lambda, no need to feed the MDC in that case)